### PR TITLE
[USER32_APTEST][USER32] CopyImage improve regression test and function.

### PR DESCRIPTION
**Pull request by Doug-Lyons**
## Purpose

_Improve CopyImage funciton by adding fast return when HANDLE is NULL and remove previous HACK._

JIRA issue: [CORE-19806](https://jira.reactos.org/browse/CORE-19806)
[CORE-17902](https://jira.reactos.org/browse/CORE-17902)

## Proposed changes

_Test for HANDLE == NULL and set LastError and do fast return if true._
_Improve previous HACK in PR #6886 and commit d3ec7cd._

### CODE REVIEW

File: win32ss/user/user32/windows/cursoricon.c [hunk 1]
```diff
@@ -2032,6 +2032,24 @@ HANDLE WINAPI CopyImage(
     TRACE("hImage=%p, uType=%u, cxDesired=%d, cyDesired=%d, fuFlags=%x\n",
         hImage, uType, cxDesired, cyDesired, fuFlags);
 
+    /* If hImage is NULL, then do fast return after setting LastError. */
```
Comment (HBelusca) [inline win32ss/user/user32/windows/cursoricon.c:2035]:
```suggestion
    /* If hImage is NULL, then do fast return after setting LastError */
```

File: win32ss/user/user32/windows/cursoricon.c [hunk 2]
```diff
@@ -2032,6 +2032,24 @@ HANDLE WINAPI CopyImage(
...
+    if (!hImage)
+    {
+        switch(uType)
```
Comment (HBelusca) [inline win32ss/user/user32/windows/cursoricon.c:2038]:
```suggestion
        switch (uType)
```

File: modules/rostests/apitests/user32/CopyImage.c [hunk 3]
```diff
@@ -76,9 +77,53 @@ Test_CopyImage_Flags(UINT uType)
...
 
+static VOID
+Test_CopyImage_hImage_NULL(void)
+{
+    HANDLE hImg;
+    DWORD LastError;
+
+    /* Test NULL HANDLE return and GetLastError return. */
+    SetLastError(0xdeadbeef);
+    hImg = CopyImage(NULL, IMAGE_ICON, 16, 16, 0x4000);
```
Comment (HBelusca) [inline modules/rostests/apitests/user32/CopyImage.c:88]:
```suggestion
    hImg = CopyImage(NULL, IMAGE_ICON, 16, 16, LR_COPYFROMRESOURCE);
```

File: modules/rostests/apitests/user32/CopyImage.c [hunk 4]
```diff
@@ -76,9 +77,53 @@ Test_CopyImage_Flags(UINT uType)
...
+    LastError = GetLastError();
+    ok(LastError == ERROR_INVALID_CURSOR_HANDLE, "Wrong error 0x%08lx returned\n", LastError);
+    ok(!hImg, "Image returned should have been NULL\n");
+
+    SetLastError(0xdeadbeef);
+    hImg = CopyImage(NULL, IMAGE_BITMAP, 16, 16, 0x4000);
```
Comment (HBelusca) [inline modules/rostests/apitests/user32/CopyImage.c:94]:
```suggestion
    hImg = CopyImage(NULL, IMAGE_BITMAP, 16, 16, LR_COPYFROMRESOURCE);
```

File: modules/rostests/apitests/user32/CopyImage.c [hunk 5]
```diff
@@ -76,9 +77,53 @@ Test_CopyImage_Flags(UINT uType)
...
+    LastError = GetLastError();
+    ok(LastError == ERROR_INVALID_HANDLE, "Wrong error 0x%08lx returned\n", LastError);
+    ok(!hImg, "Image returned should have been NULL\n");
+
+
+    SetLastError(0xdeadbeef);
+    hImg = CopyImage(NULL, IMAGE_CURSOR, 16, 16, 0x4000);
```
Comment (HBelusca) [inline modules/rostests/apitests/user32/CopyImage.c:101]:
```suggestion
    hImg = CopyImage(NULL, IMAGE_CURSOR, 16, 16, LR_COPYFROMRESOURCE);
```

File: modules/rostests/apitests/user32/CopyImage.c [hunk 6]
```diff
@@ -76,9 +77,53 @@ Test_CopyImage_Flags(UINT uType)
...
+    LastError = GetLastError();
+    ok(LastError == ERROR_INVALID_CURSOR_HANDLE, "Wrong error 0x%08lx returned\n", LastError);
+    ok(!hImg, "Image returned should have been NULL\n");
+
+    /* Test bad Type (5) GetLastError return value. Not Icon, Cursor, or Bitmap. */
+    SetLastError(0xdeadbeef);
+    hImg = CopyImage(NULL, 5, 16, 16, 0x4000);
```
Comment (HBelusca) [inline modules/rostests/apitests/user32/CopyImage.c:108]:
```suggestion
    hImg = CopyImage(NULL, 5, 16, 16, LR_COPYFROMRESOURCE);
```

File: modules/rostests/apitests/user32/CopyImage.c [hunk 7]
```diff
@@ -76,9 +77,53 @@ Test_CopyImage_Flags(UINT uType)
...
+    LastError = GetLastError();
+    ok(LastError == ERROR_INVALID_PARAMETER, "Wrong error 0x%08lx returned\n", LastError);
+    ok(!hImg, "Image returned should have been NULL\n");
+
+    /* Test bad type (5) GetLastError return value with good HANDLE */
+    hImg = CreateTestImage(IMAGE_ICON);
+    SetLastError(0xdeadbeef);
+    hImg = CopyImage(hImg, 5, 16, 16, 0x4000);
```
Comment (HBelusca) [inline modules/rostests/apitests/user32/CopyImage.c:116]:
```suggestion
    hImg = CopyImage(hImg, 5, 16, 16, LR_COPYFROMRESOURCE);
```

File: win32ss/user/user32/windows/cursoricon.c [hunk 8]
```diff
@@ -2032,6 +2032,24 @@ HANDLE WINAPI CopyImage(
...
+            case IMAGE_ICON:
+                SetLastError(ERROR_INVALID_CURSOR_HANDLE);
+                break;
+            default:
+                SetLastError(ERROR_INVALID_PARAMETER);
+        }
+        return NULL;
+    }
+
     if (fuFlags & ~COPYIMAGE_VALID_FLAGS)
```
Comment (HBelusca) [inline win32ss/user/user32/windows/cursoricon.c:2053]: Can you also test the following:
```c
CopyImage(NULL, IMAGE_BITMAP, 16, 16, 0x80000000);
```
(`0x80000000` is an invalid flag value), in order to test whether the flags validation is done first (we would get an `ERROR_INVALID_PARAMETER`), or whether the handle NULL check is done first instead, as you suggest in your PR (we would get an `ERROR_INVALID_HANDLE`).

File: win32ss/user/user32/windows/cursoricon.c [hunk 9]
```diff
@@ -2044,14 +2062,39 @@ HANDLE WINAPI CopyImage(
...
+                DeleteObject(iconinfo.hbmColor);
+                ReleaseDC(NULL, hDC);
+
+                /* If the images are the same size remove LF_COPYFROMRESOURCE and try again. */
+                if (cxDesired == bmi.bmiHeader.biWidth && cyDesired == bmi.bmiHeader.biHeight)
+                    handle = CURSORICON_CopyImage(hImage, uType == IMAGE_ICON, cxDesired,
+                                                  cyDesired, (fuFlags & ~LR_COPYFROMRESOURCE));
+            }
+            if (handle)
+                return handle;
```
Comment (HBelusca) [inline win32ss/user/user32/windows/cursoricon.c:2097]:
```suggestion
            return handle;
```
If the handle is not NULL, this behaves the same, and if the handle is NULL, it's going to do the same as the `return NULL;` below.

File: win32ss/user/user32/windows/cursoricon.c [hunk 10]
```diff
@@ -2044,14 +2062,39 @@ HANDLE WINAPI CopyImage(
...
+                struct
+                {
+                    BITMAPINFOHEADER bmiHeader;
+                    RGBQUAD bmiColors[256];
+                } bmi;
+
+                if (!GetIconInfo((HICON)hImage, &iconinfo))
+                    ERR("GetIconInfo Failed. hImage %p\n", hImage);
+                memset(&bmi, 0, sizeof(bmi));
+                bmi.bmiHeader.biSize = sizeof(BITMAPINFOHEADER);
```
Comment (HBelusca) [inline win32ss/user/user32/windows/cursoricon.c:2083]:
```suggestion
                ZeroMemory(&bmi, sizeof(bmi));
                bmi.bmiHeader.biSize = sizeof(bmi.bmiHeader);
```

File: win32ss/user/user32/windows/cursoricon.c [hunk 11]
```diff
@@ -2044,14 +2062,39 @@ HANDLE WINAPI CopyImage(
...
+                hDC = GetDC(NULL);  // Screen hdc
+                GetDIBits(hDC, iconinfo.hbmMask, 0, 1, NULL, (BITMAPINFO *)&bmi, DIB_RGB_COLORS);
+                DeleteObject(iconinfo.hbmMask);
+                DeleteObject(iconinfo.hbmColor);
+                ReleaseDC(NULL, hDC);
+
+                /* If the images are the same size remove LF_COPYFROMRESOURCE and try again. */
+                if (cxDesired == bmi.bmiHeader.biWidth && cyDesired == bmi.bmiHeader.biHeight)
+                    handle = CURSORICON_CopyImage(hImage, uType == IMAGE_ICON, cxDesired,
+                                                  cyDesired, (fuFlags & ~LR_COPYFROMRESOURCE));
```
Comment (HBelusca) [inline win32ss/user/user32/windows/cursoricon.c:2094]:
```suggestion
                if (cxDesired == bmi.bmiHeader.biWidth && cyDesired == bmi.bmiHeader.biHeight)
                {
                    handle = CURSORICON_CopyImage(hImage, uType == IMAGE_ICON, cxDesired,
                                                  cyDesired, (fuFlags & ~LR_COPYFROMRESOURCE));
                }
```

File: win32ss/user/user32/windows/cursoricon.c (hunk 8 above)
Comment (Doug-Lyons) [reply win32ss/user/user32/windows/cursoricon.c]: I added another test as you requested and it showed that with the ```0x80000000``` invalid flag value and the NULL handle, the LastError returned was ```ERROR_INVALID_PARAMETER```. Therefore, I moved the HANDLE test after the Invalid Flags test in the CopyImage function. That was a good idea, @HBelusca. Thanks.

File: win32ss/user/user32/windows/cursoricon.c (hunk 1 above)
Comment (Doug-Lyons) [reply win32ss/user/user32/windows/cursoricon.c]: Changed. Thanks.

File: win32ss/user/user32/windows/cursoricon.c (hunk 2 above)
Comment (Doug-Lyons) [reply win32ss/user/user32/windows/cursoricon.c]: Changed. Thanks.

File: modules/rostests/apitests/user32/CopyImage.c (hunk 3 above)
Comment (Doug-Lyons) [reply modules/rostests/apitests/user32/CopyImage.c]: Changed. Thanks.

File: modules/rostests/apitests/user32/CopyImage.c (hunk 4 above)
Comment (Doug-Lyons) [reply modules/rostests/apitests/user32/CopyImage.c]: Changed. Thanks.

File: modules/rostests/apitests/user32/CopyImage.c (hunk 5 above)
Comment (Doug-Lyons) [reply modules/rostests/apitests/user32/CopyImage.c]: Changed. Thanks.

File: modules/rostests/apitests/user32/CopyImage.c (hunk 6 above)
Comment (Doug-Lyons) [reply modules/rostests/apitests/user32/CopyImage.c]: Changed. Thanks.

File: modules/rostests/apitests/user32/CopyImage.c (hunk 7 above)
Comment (Doug-Lyons) [reply modules/rostests/apitests/user32/CopyImage.c]: Changed. Thanks.

File: win32ss/user/user32/windows/cursoricon.c (hunk 9 above)
Comment (Doug-Lyons) [reply win32ss/user/user32/windows/cursoricon.c]: Changed. Thanks.

File: win32ss/user/user32/windows/cursoricon.c (hunk 10 above)
Comment (Doug-Lyons) [reply win32ss/user/user32/windows/cursoricon.c]: Changed. Thanks.

File: win32ss/user/user32/windows/cursoricon.c (hunk 11 above)
Comment (Doug-Lyons) [reply win32ss/user/user32/windows/cursoricon.c]: Changed. Thanks.

File: win32ss/user/user32/windows/cursoricon.c [hunk 12]
```diff
@@ -2038,19 +2038,63 @@ HANDLE WINAPI CopyImage(
...
+                break;
+            case IMAGE_CURSOR:
+            case IMAGE_ICON:
+                SetLastError(ERROR_INVALID_CURSOR_HANDLE);
+                break;
+            default:
+                SetLastError(ERROR_INVALID_PARAMETER);
+        }
+        return NULL;
+    }
```
Comment (HBelusca) [inline win32ss/user/user32/windows/cursoricon.c:2057]: Since now these two parts of code are successive, I would suggest you just incorporate inside the existing switch below these null-tests, see the following suggestions.
The default case that returns `ERROR_INVALID_PARAMETER` is already handled in the other switch.

File: win32ss/user/user32/windows/cursoricon.c [hunk 13]
```diff
@@ -2038,19 +2038,63 @@ HANDLE WINAPI CopyImage(
...
+
     switch(uType)
     {
         case IMAGE_BITMAP:
             return BITMAP_CopyImage(hImage, cxDesired, cyDesired, fuFlags);
```
Comment (HBelusca) [inline win32ss/user/user32/windows/cursoricon.c:2062]:
```suggestion
            if (!hImage)
            {
                SetLastError(ERROR_INVALID_HANDLE);
                break;
            }
            return BITMAP_CopyImage(hImage, cxDesired, cyDesired, fuFlags);
```
(seeing this I'm also wondering whether the NULL-check + last-error setting could be included within the BITMAP_CopyImage function itself.)

File: win32ss/user/user32/windows/cursoricon.c [hunk 14]
```diff
@@ -2038,19 +2038,63 @@ HANDLE WINAPI CopyImage(
...
         case IMAGE_CURSOR:
         case IMAGE_ICON:
-        /* HACK: Copying bitmaps with LR_COPYFROMRESOURCE flag fails. CORE-17902.
-         * This is a way to return the original bit map if we need
-         * the icons to show up. We need a simpler test. */
         {
             HANDLE handle = CURSORICON_CopyImage(hImage, uType == IMAGE_ICON, cxDesired, cyDesired, fuFlags);
```
Comment (HBelusca) [inline win32ss/user/user32/windows/cursoricon.c:2066]:
```suggestion
            HANDLE handle;
            if (!hImage)
            {
                SetLastError(ERROR_INVALID_CURSOR_HANDLE);
                break;
            }
            handle = CURSORICON_CopyImage(hImage, uType == IMAGE_ICON, cxDesired, cyDesired, fuFlags);
```
(Again, I'm wondering whether the NULL-check + last-error setting could be included within the CURSORICON_CopyImagefunction itself. Notice that the CopyIcon() function found later below might benefit from this, if the last-error check turns out to be the same.)

File: win32ss/user/user32/windows/cursoricon.c (hunk 13 above)
Comment (Doug-Lyons) [reply win32ss/user/user32/windows/cursoricon.c:2049]: Changed. Thanks.

File: win32ss/user/user32/windows/cursoricon.c (hunk 14 above)
Comment (Doug-Lyons) [reply win32ss/user/user32/windows/cursoricon.c]: Changed. Thanks. For the CURSORICON_CopyImage function comments, maybe this can be considered in another PR.

File: win32ss/user/user32/windows/cursoricon.c [hunk 15]
```diff
@@ -2041,16 +2041,53 @@ HANDLE WINAPI CopyImage(
...
+                ZeroMemory(&bmi, sizeof(bmi));
+                bmi.bmiHeader.biSize = sizeof(bmi.bmiHeader);
+
+                hDC = GetDC(NULL);  // Screen hdc
+                GetDIBits(hDC, iconinfo.hbmMask, 0, 1, NULL, (BITMAPINFO *)&bmi, DIB_RGB_COLORS);
+                DeleteObject(iconinfo.hbmMask);
+                DeleteObject(iconinfo.hbmColor);
+                ReleaseDC(NULL, hDC);
+
+                /* If the images are the same size remove LF_COPYFROMRESOURCE and try again. */
```
Comment (HBelusca) [inline win32ss/user/user32/windows/cursoricon.c:2084]:
```suggestion
                /* If the images are the same size remove LF_COPYFROMRESOURCE and try again */
```

File: win32ss/user/user32/windows/cursoricon.c [hunk 16]
```diff
@@ -2041,16 +2041,53 @@ HANDLE WINAPI CopyImage(
...
+                ICONINFO iconinfo = { 0 };
+                /* struct is for BITMAPINFO plus a full-sized color table */
+                struct
+                {
+                    BITMAPINFOHEADER bmiHeader;
+                    RGBQUAD bmiColors[256];
+                } bmi;
+
+                if (!GetIconInfo((HICON)hImage, &iconinfo))
+                    ERR("GetIconInfo Failed. hImage %p\n", hImage);
```
Comment (HBelusca) [inline win32ss/user/user32/windows/cursoricon.c:2074]: should we continue if this call fails? (the iconinfo.hbmMask and hbmColor below would remain NULL in that case)

File: win32ss/user/user32/windows/cursoricon.c [hunk 17]
```diff
@@ -2041,16 +2041,53 @@ HANDLE WINAPI CopyImage(
...
+                ZeroMemory(&bmi, sizeof(bmi));
+                bmi.bmiHeader.biSize = sizeof(bmi.bmiHeader);
+
+                hDC = GetDC(NULL);  // Screen hdc
+                GetDIBits(hDC, iconinfo.hbmMask, 0, 1, NULL, (BITMAPINFO *)&bmi, DIB_RGB_COLORS);
```
Comment (HBelusca) [inline win32ss/user/user32/windows/cursoricon.c:2079]:
```suggestion
                GetDIBits(hDC, iconinfo.hbmMask, 0, 1, NULL, (PBITMAPINFO)&bmi, DIB_RGB_COLORS);
```

File: win32ss/user/user32/windows/cursoricon.c [hunk 18]
```diff
@@ -2041,16 +2041,53 @@ HANDLE WINAPI CopyImage(
...
+                    BITMAPINFOHEADER bmiHeader;
+                    RGBQUAD bmiColors[256];
+                } bmi;
+
+                if (!GetIconInfo((HICON)hImage, &iconinfo))
+                    ERR("GetIconInfo Failed. hImage %p\n", hImage);
+                ZeroMemory(&bmi, sizeof(bmi));
+                bmi.bmiHeader.biSize = sizeof(bmi.bmiHeader);
+
+                hDC = GetDC(NULL);  // Screen hdc
```
Comment (HBelusca) [inline win32ss/user/user32/windows/cursoricon.c:2078]: In principle this call shouldn't fail, but....

**Review by HBelusca — APPROVED** [review]

File: win32ss/user/user32/windows/cursoricon.c (hunk 16 above)
Comment (Doug-Lyons) [reply win32ss/user/user32/windows/cursoricon.c:2075]: I have revised the code so that on an error here, we do ```return NULL;```. Thanks

File: win32ss/user/user32/windows/cursoricon.c (hunk 18 above)
Comment (Doug-Lyons) [reply win32ss/user/user32/windows/cursoricon.c:2081]: @HBelusca, I did a non-exhaustive search of the ReactOS code base and in most cases where we use this code:
```
GetDC(NULL);
```
there is no error checking, but since you pointed this out, now I check if ```hDC``` is NULL.
If so, we will output an error message and skip the GetDIBits call that used this value.
We needed to free the ```iconinfo``` items, so we could not return here, but eventually after the failing "IF" test below, we will return "NULL".

@whindsaks, Sorry I do not understand your question about DIB_RGB_COLORS. Can you be more specific. Thanks.

File: win32ss/user/user32/windows/cursoricon.c (hunk 15 above)
Comment (Doug-Lyons) [reply win32ss/user/user32/windows/cursoricon.c]: Changed. Thanks.

File: win32ss/user/user32/windows/cursoricon.c (hunk 17 above)
Comment (Doug-Lyons) [reply win32ss/user/user32/windows/cursoricon.c]: Changed. Thanks.

File: win32ss/user/user32/windows/cursoricon.c (hunk 18 above)
Comment (whindsaks) [reply win32ss/user/user32/windows/cursoricon.c:2081]: My question was, does GetDIBits need a hdc with DIB_RGB_COLORS or is it only really needed for DIB_PAL_COLORS (palette).

Comment (Doug-Lyons) [reply win32ss/user/user32/windows/cursoricon.c]: @whindsaks, I do not really know the answer to your question, but it was a good one. In researching it, I found a much simpler way to get the width and height from an HBITMAP. So instead of using GetDIBits, I now use GetObject. So there is no longer any  use of the DIB_PAL_COLORS of DIB_RGB_COLORS involved here. Thanks.

**Review by oleg-dubinskiy — APPROVED** [review]

**Review by DarkFire01 — APPROVED** [review]

**Comment by Doug-Lyons** [comment]
Thanks for everyone's help with this and for approvals from @HBelusca, @oleg-dubinskiy and @DarkFire01.
